runtime-sdk/modules/rofl-market/src/types.rs

```diff
@@ -0,0 +1,518 @@
+#[cfg_attr(test, derive(PartialEq, Eq))]
+pub struct GpuResource {
+    /// Optional GPU model identifier.
+    pub model: Option<String>,
+    /// Number of requested GPUs.
+    pub count: u8,
+}
+
+/// Provisioned instance descriptor.
+#[derive(Clone, Debug, Default, cbor::Encode, cbor::Decode)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+pub struct Instance {
+    /// Address of the provider.
+    pub provider: Address,
+    /// Per-provider unique instance identifier.
+    pub id: InstanceId,
+    /// Per-provider offer identifier.
+    pub offer: OfferId,
+    /// Status of the instance.
+    pub status: InstanceStatus,
+    /// Address of the administrator account.
+    pub admin: Address,
+    /// Identifier of the node where the instance has been provisioned.
+    pub node_id: Option<PublicKey>,
+    /// Arbitrary metadata (key-value pairs) assigned by the provider's scheduler.
+    pub metadata: BTreeMap<String, String>,
+    /// Deployed instance resources.
+    pub resources: Resources,
+    /// Current deployment running on this instance.
+    pub deployment: Option<Deployment>,
+    /// Timestamp when the instance was created at.
+    pub created_at: u64,
+    /// Timestamp when the instance was last updated at.
+    pub updated_at: u64,
+
+    /// Timestamp from which the instance has been paid for and not yet claimed by the provider.
+    pub paid_from: u64,
+    /// Timestamp until which the instance has been paid for.
+    pub paid_until: u64,
+    /// Payment information for this instance (copied from offer so that we can handle top-ups and
+    /// refunds even when the provider changes the original offers).
+    pub payment: Payment,
+    /// Instance payment address.
+    pub payment_address: [u8; 20],
+    /// Payment method-specific refund information.
+    pub refund_data: Vec<u8>,
+
+    /// Next command identifier to use.
+    pub cmd_next_id: CommandId,
+    /// Number of queued commands.
+    pub cmd_count: u64,
+}
+
+impl Instance {
+    /// Whether the instance has been accepted by the provider.
+    pub fn is_accepted(&self) -> bool {
+        matches!(self.status, InstanceStatus::Accepted)
+    }
+}
+
+/// Instance status.
+#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, Hash, cbor::Encode, cbor::Decode)]
+#[cbor(with_default)]
+#[repr(u8)]
+pub enum InstanceStatus {
+    /// Instance is pending to be accepted.
+    #[default]
+    Created = 0,
+
+    /// Instance has been accepted by the provider.
+    Accepted = 1,
+
+    /// Instance has been cancelled.
+    Cancelled = 2,
+}
+
+/// Descriptor of what is deployed into an instance.
+#[derive(Clone, Debug, Default, cbor::Encode, cbor::Decode)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+pub struct Deployment {
+    /// Identifier of the deployed ROFL app.
+    pub app_id: AppId,
+    /// ROFL app manifest hash.
+    pub manifest_hash: Hash,
+    /// Arbitrary metadata (key-value pairs) assigned by the deployer.
+    pub metadata: BTreeMap<String, String>,
```

Comment:
Should we add some validation to limit the length to a reasonable value? The same applies to other metadata key-value pairs and even the existing metadata/secret key-value pairs in the ROFL module.

I guess providing very large values would increase the transaction sizes which provides some minimal protection.

Reply:
Fixed.